File: utils/file_handler.py
import os
import json
import boto3
from typing import Any, Dict, Optional
from botocore.exceptions import ClientError, NoCredentialsError
import sys
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from config import config

class FileHandler:
    """Handles file operations for both local and S3 storage"""
    
    def __init__(self):
        self.s3_client = None
        if config.IS_PRODUCTION and config.AWS_ACCESS_KEY_ID:
            try:
                self.s3_client = boto3.client(
                    's3',
                    aws_access_key_id=config.AWS_ACCESS_KEY_ID,
                    aws_secret_access_key=config.AWS_SECRET_ACCESS_KEY,
                    region_name=config.AWS_REGION
                )
            except NoCredentialsError:
                print("Warning: AWS credentials not found, falling back to local storage")
                self.s3_client = None
    
    def is_s3_path(self, path: str) -> bool:
        """Check if a path is an S3 path"""
        return path.startswith('s3://')
    
    def read_json(self, path: str) -> Optional[Dict[str, Any]]:
        """Read JSON data from local file or S3"""
        try:
            if self.is_s3_path(path):
                return self._read_json_from_s3(path)
            else:
                return self._read_json_from_local(path)
        except Exception as e:
            print(f"Error reading JSON from {path}: {e}")
            return None
    
    def write_json(self, data: Dict[str, Any], path: str) -> bool:
        """Write JSON data to local file or S3"""
        try:
            if self.is_s3_path(path):
                return self._write_json_to_s3(data, path)
            else:
                return self._write_json_to_local(data, path)
        except Exception as e:
            print(f"Error writing JSON to {path}: {e}")
            return False
    
    def file_exists(self, path: str) -> bool:
        """Check if file exists in local filesystem or S3"""
        try:
            if self.is_s3_path(path):
                return self._s3_file_exists(path)
            else:
                return os.path.exists(path)
        except Exception as e:
            print(f"Error checking file existence for {path}: {e}")
            return False
    
    def open_binary(self, path: str):
        """Open binary file for reading (local only for now)"""
        if self.is_s3_path(path):
            raise NotImplementedError("Binary file reading from S3 not yet implemented")
        else:
            return open(path, 'rb')
    
    def _read_json_from_local(self, path: str) -> Dict[str, Any]:
        """Read JSON from local file"""
        with open(path, 'r', encoding='utf-8') as f:
            return json.load(f)
    
    def _write_json_to_local(self, data: Dict[str, Any], path: str) -> bool:
        """Write JSON to local file"""
        # Ensure directory exists
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        return True
    
    def _read_json_from_s3(self, s3_path: str) -> Dict[str, Any]:
        """Read JSON from S3"""
        if not self.s3_client:
            raise Exception("S3 client not initialized")
        
        bucket, key = self._parse_s3_path(s3_path)
        response = self.s3_client.get_object(Bucket=bucket, Key=key)
        return json.loads(response['Body'].read().decode('utf-8'))
    
    def _write_json_to_s3(self, data: Dict[str, Any], s3_path: str) -> bool:
        """Write JSON to S3"""
        if not self.s3_client:
            raise Exception("S3 client not initialized")
        
        bucket, key = self._parse_s3_path(s3_path)
        json_data = json.dumps(data, indent=2, ensure_ascii=False)
        self.s3_client.put_object(
            Bucket=bucket,
            Key=key,
            Body=json_data.encode('utf-8'),
            ContentType='application/json'
        )
        return True
    
    def _s3_file_exists(self, s3_path: str) -> bool:
        """Check if file exists in S3"""
        if not self.s3_client:
            return False
        
        bucket, key = self._parse_s3_path(s3_path)
        try:
            self.s3_client.head_object(Bucket=bucket, Key=key)
            return True
        except ClientError as e:
            if e.response['Error']['Code'] == '404':
                return False
            raise
    
    def _parse_s3_path(self, s3_path: str) -> tuple[str, str]:
        """Parse S3 path to get bucket and key"""
        if not s3_path.startswith('s3://'):
            raise ValueError(f"Invalid S3 path: {s3_path}")
        
        path_parts = s3_path[5:].split('/', 1)
        if len(path_parts) != 2:
            raise ValueError(f"Invalid S3 path format: {s3_path}")
        
        return path_parts[0], path_parts[1]

# Global file handler instance
file_handler = FileHandler() 

File: referralguard-dashboard/lib/file-handler.ts
import { S3Client, GetObjectCommand, PutObjectCommand, HeadObjectCommand } from '@aws-sdk/client-s3';
import { config } from './config';

export interface FileHandlerInterface {
  isS3Path(path: string): boolean;
  readJson(path: string): Promise<any | null>;
  writeJson(data: any, path: string): Promise<boolean>;
  fileExists(path: string): Promise<boolean>;
}

export class FileHandler implements FileHandlerInterface {
  private s3Client: S3Client | null = null;

  constructor() {
    if (config.isProduction && config.awsCredentials.accessKeyId) {
      try {
        this.s3Client = new S3Client({
          region: config.awsCredentials.region,
          credentials: {
            accessKeyId: config.awsCredentials.accessKeyId,
            secretAccessKey: config.awsCredentials.secretAccessKey!
          }
        });
      } catch (error) {
        console.warn("Warning: AWS credentials not found, falling back to local storage");
        this.s3Client = null;
      }
    }
  }

  isS3Path(path: string): boolean {
    return path.startsWith('s3://');
  }

  async readJson(path: string): Promise<any | null> {
    try {
      if (this.isS3Path(path)) {
        return await this.readJsonFromS3(path);
      } else {
        return await this.readJsonFromLocal(path);
      }
    } catch (error) {
      console.error(`Error reading JSON from ${path}:`, error);
      return null;
    }
  }

  async writeJson(data: any, path: string): Promise<boolean> {
    try {
      if (this.isS3Path(path)) {
        return await this.writeJsonToS3(data, path);
      } else {
        return await this.writeJsonToLocal(data, path);
      }
    } catch (error) {
      console.error(`Error writing JSON to ${path}:`, error);
      return false;
    }
  }

  async fileExists(path: string): Promise<boolean> {
    try {
      if (this.isS3Path(path)) {
        return await this.s3FileExists(path);
      } else {
        // For local files, we'll check if the file exists in the public directory
        // This is a simplified approach for Next.js
        return true; // We'll handle the actual file reading in readJsonFromLocal
      }
    } catch (error) {
      console.error(`Error checking file existence for ${path}:`, error);
      return false;
    }
  }

  private async readJsonFromLocal(path: string): Promise<any> {
    // For Next.js, we'll read from the public directory
    const response = await fetch(`/api/file?path=${encodeURIComponent(path)}`);
    if (!response.ok) {
      throw new Error(`Failed to read file: ${response.statusText}`);
    }
    return await response.json();
  }

  private async writeJsonToLocal(data: any, path: string): Promise<boolean> {
    // For Next.js, we'll write to the public directory via API
    const response = await fetch('/api/file', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ path, data }),
    });
    return response.ok;
  }

  private async readJsonFromS3(s3Path: string): Promise<any> {
    if (!this.s3Client) {
      throw new Error("S3 client not initialized");
    }

    const { bucket, key } = this.parseS3Path(s3Path);
    const command = new GetObjectCommand({ Bucket: bucket, Key: key });
    const response = await this.s3Client.send(command);
    
    if (!response.Body) {
      throw new Error("Empty response body from S3");
    }

    const text = await response.Body.transformToString();
    return JSON.parse(text);
  }

  private async writeJsonToS3(data: any, s3Path: string): Promise<boolean> {
    if (!this.s3Client) {
      throw new Error("S3 client not initialized");
    }

    const { bucket, key } = this.parseS3Path(s3Path);
    const jsonData = JSON.stringify(data, null, 2);
    const command = new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: jsonData,
      ContentType: 'application/json'
    });

    await this.s3Client.send(command);
    return true;
  }

  private async s3FileExists(s3Path: string): Promise<boolean> {
    if (!this.s3Client) {
      return false;
    }

    const { bucket, key } = this.parseS3Path(s3Path);
    try {
      const command = new HeadObjectCommand({ Bucket: bucket, Key: key });
      await this.s3Client.send(command);
      return true;
    } catch (error: any) {
      if (error.name === 'NotFound') {
        return false;
      }
      throw error;
    }
  }

  private parseS3Path(s3Path: string): { bucket: string; key: string } {
    if (!s3Path.startsWith('s3://')) {
      throw new Error(`Invalid S3 path: ${s3Path}`);
    }

    const pathParts = s3Path.slice(5).split('/', 1);
    if (pathParts.length !== 2) {
      throw new Error(`Invalid S3 path format: ${s3Path}`);
    }

    return { bucket: pathParts[0], key: pathParts[1] };
  }
}

export const fileHandler = new FileHandler(); 